pkg/builtins/signature.go

```diff
@@ -0,0 +1,100 @@
+package builtins
+
+import (
+	"github.com/lambdaclass/cairo-vm.go/pkg/lambdaworks"
+	"github.com/lambdaclass/cairo-vm.go/pkg/starknet_crypto"
+	"github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+	"github.com/pkg/errors"
+)
+
+const SIGNATURE_BUILTIN_NAME = "signature"
+
+// Notice changing this to any other number breaks the code
+const SIGNATURE_CELLS_PER_INSTANCE = 2
+
+type Signature struct {
+	R lambdaworks.Felt
+	S lambdaworks.Felt
+}
+type SignatureBuiltinRunner struct {
+	base       memory.Relocatable
+	included   bool
+	signatures map[memory.Relocatable](Signature)
+}
+
+func (signatureRunner *SignatureBuiltinRunner) Base() memory.Relocatable {
+	return signatureRunner.base
+}
+
+func (signatureRunner *SignatureBuiltinRunner) Name() string {
+	return SIGNATURE_BUILTIN_NAME
+}
+
+func (signatureRunner *SignatureBuiltinRunner) InitializeSegments(segments *memory.MemorySegmentManager) {
+	signatureRunner.base = segments.AddSegment()
+}
+
+func (signatureRunner *SignatureBuiltinRunner) InitialStack() []memory.MaybeRelocatable {
+	if signatureRunner.included {
+		return []memory.MaybeRelocatable{*memory.NewMaybeRelocatableRelocatable(signatureRunner.base)}
+	} else {
+		return nil
+	}
+}
+
+func (signatureRunner *SignatureBuiltinRunner) DeduceMemoryCell(addr memory.Relocatable, mem *memory.Memory) (*memory.MaybeRelocatable, error) {
+	return nil, nil
+}
+
+func SignatureVerificationError() error {
+	return errors.New("Signature is not valid")
+}
+
+func (r *SignatureBuiltinRunner) Include(include bool) {
+	r.included = include
+}
+
+func ValidationRuleSignature(mem *memory.Memory, address memory.Relocatable, signatureBuiltin *SignatureBuiltinRunner) ([]memory.Relocatable, error) {
+	cell_index := address.Offset % SIGNATURE_CELLS_PER_INSTANCE
+	var pub_key_address, message_addr memory.Relocatable
+
+	if cell_index == 0 {
+		pub_key_address = address
+		message_addr = address.AddUint(1)
+		// This should be 1, since CELLS_PER_INSTANCE is 2.
+	} else {
+		pub_key_address, _ = address.SubUint(1)
+		message_addr = address
+	}
+
+	pub_key, get_pubkey_error := mem.GetFelt(pub_key_address)
+	message, get_message_error := mem.GetFelt(message_addr)
+	signature, found_signature := signatureBuiltin.signatures[pub_key_address]
+
+	if !found_signature || get_pubkey_error != nil || get_message_error != nil {
+		return nil, SignatureVerificationError()
+	}
```

Comment:
This differs from the canonic implementation
The cairo-vm does the following:
```
let pubkey = match memory.get_integer(pubkey_addr) {
                    Ok(num) => num,
                    Err(_) if cell_index == 1 => return Ok(vec![]),
                    _ => return Err(MemoryError::PubKeyNonInt(Box::new(pubkey_addr))),
                };

                let msg = match memory.get_integer(message_addr) {
                    Ok(num) => num,
                    Err(_) if cell_index == 0 => return Ok(vec![]),
                    _ => return Err(MemoryError::MsgNonInt(Box::new(message_addr))),
                };
```
This is so that if an input set (address & msg cells) is incomplete, then the validation is not ran. This makes sense, as values are written into memory once at a time. For example, If I were to write the pubkey and then the msg into memory, when writing the pubkey this validation will run, and we don't want to shut down the vm in that case, we want to skip this case and then run the validation when I write the msg and the input set of cells is complete

Reply:
There it should be fixed